BUG: improve handling of Series.value_counts's argument 'dropna' (GH9443)

Kodiologist opened the pull request:
closes #9443

jreback commented on pandas/core/algorithms.py:
```diff
@@ -202,8 +202,13 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
+    values = Series(values)
     is_category = com.is_categorical_dtype(values.dtype)
+    if is_category and dropna:
+        values = values.dropna()
```
this is better to dispatch to the Categorical (eg it's the values) and define dropna() there

jreback commented on pandas/core/algorithms.py:
```diff
@@ -240,18 +245,25 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
-            result.index = bins[:-1]
+        if is_category:
+            ix = np.arange(len(cat.categories))
+            if not dropna and -1 in values:
```
we do something slmost exactly like this inside Categorical do may need to pull that out as a method and call it here (or maybe remove_categories can suffice)

Kodiologist replied:
Are you thinking of `Categorical.describe`? It puts the NaN category at the end rather than the beginning. There doesn't seem to be much to factor out, although `Categorical.describe` does basically the same thing as `Series.value_counts`.

jreback replied:
what I mean is that this entire section should be internal to a Categorical. Define a ``value_counts`` in Categorical. and just call it here (yes ``.describe()`` does do some of this, you might be able to refactor some of this code out into a private method to avoid duplication).

Kodiologist replied:
Is `value_counts` a good name for this new method of Categorical if all it does is fix the index after the value-counts series has already been built (in this case, by the line `result = Series(counts, index=com._values_from_object(keys))` above)? Or do you mean that most or all of the Categorical-related code in `algorithms.value_counts` should be moved to this new method?

jreback replied:
it should be what the value_count part of describe does. (the groupby)

Kodiologist replied:
Sorry for being dense, but I think I still don't understand. The `.groupby` line in `Categorical.describe` only makes a table of counts of the raw codes. It doesn't fix the index, which is what this diff here is about.

jreback replied:
you can make a ``Categorical.value_counts(dropna=False)`` method to do exactly this which you will then call in the ``core/algorithms.py``. (and of course it should handle ``dropna=True`` as well.

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -254,6 +254,21 @@ def test_value_counts_nat(self):
         # TODO same for (timedelta)
 
+    def test_dropna(self):
+        tm.assert_series_equal(
```
pls add a test that has nan in it as well (for these dtypes too, e.g. bool, float), even though adding nan will make the bool -> object

jreback replied:
add the issue number as a comment at the top of the test

Kodiologist commented:
If `Categorical.dropna` looks good, I could add tests for it, too, of course.

jreback commented on pandas/core/algorithms.py:
```diff
@@ -204,6 +204,10 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
     is_category = com.is_categorical_dtype(values.dtype)
+    if is_category and dropna:
+        values = values.dropna()
+        if np.nan in values.categories:
```
you shouldn't try to do ``np.nan`` in anything....its not generally a safe operation. the ``.dropna()`` should handle this completely.

Kodiologist replied:
Should `Categorical.dropna` use `np.nan in self.categories`, or is there a better way to do it?

jreback replied:
you could do

isnull(self.categories).any()

jreback commented on pandas/core/categorical.py:
```diff
@@ -881,6 +881,25 @@ def notnull(self):
+        """
+        Return the Categorical without null values.
+
+        Both missing values (-1 in .codes) and NA as a category are detected.
```
by definition this should remove nan from the categories (that's the point of the method). I don't like inplace in general and not necessary here, pls take it out. ``_update_inplace`` doesn't exist its a generic method. (so clearly this is not tested).

jorisvandenbossche commented:
Is it needed that `dropna` and `value_counts` are added as public functions to `Categorical`? 
In my ming, this goes a bit against the reasoning of keeping `Categorical` an array like (as the 'real' numpy array does not support such dtype).

jreback commented:
@jorisvandenbossche I don't see any problem with having these functions. They can be public or not, doesn't make much difference. It does not matter that numpy doesn't support these. It likely never will.

``Categorical`` acts array-like, but so what.

Kodiologist commented:
@jreback, are you waiting for anything else from me, or are we good to go?

jreback commented on pandas/core/categorical.py:
```diff
@@ -881,6 +881,54 @@ def notnull(self):
+            else cat.categories)
+
+        return result
+
```
I'd like you to fix ``Categorical.describe`` to now use this. The current impl is a hack which does basically this.

Kodiologist replied:
Okay. Currently, `Categorical.describe` puts the NA row at the end, whereas `.value_counts` puts it at the beginning. Should that be changed or left as is?

jreback replied:
these should be consistent. We currently have a ``na_position`` argument to ``.order()``. But I don't think this is necessary here. Let's make it ``last`` for both. 
@JanSchulz ?

Kodiologist replied:
Shall I interpret @JanSchulz's silence as agreement?

JanSchulz replied:
I've no opinion here :-)

Kodiologist replied:
Okay, I think I can do this later today or tomorrow, then.

jreback commented on doc/source/whatsnew/v0.16.0.txt:
give a better explanation here

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -20,6 +20,8 @@ users upgrade to this version.
 New features
 ~~~~~~~~~~~~
 
+- Added ``Categorical.value_counts`` and ``Categorical.dropna``
```
add the issue number

jreback commented:
@jorisvandenbossche 
cc @Janschulz

so adding ``.value_counts/.dropna`` to ``Categorical``, really to make the calling code simpler. Question is do we / should we expose these as public-ish method, ala ``.cat`` ?

jreback commented:
@Kodiologist pls add ``.value_counts/.dropna`` to API.rst (in Categorical section).

JanSchulz commented:
@jreback, @Kodiologist: I think these two methods should neither be added to `Series.cat` nor to api.rst: the categorical section in pai.rst is on purpose only the methods under `Series.cat` and these methods are all categorical specific. In this case it is simple, that implementation of a method in Series is different for `Categorical` than `ndarray`. During the initial iterations, many methods from Categorical, which I had added to api.rst, were removed in the end for that reason: min,max,... 

I think public is fine, as long as the API documentation makes it clear that using `Categorical` directly or `s.values` is prone to break over the time. If these should become private, it makes sense to also mark all other methods (min, max,...) private and I don't think that's possible (some are used as a drop-in for public `ndarray` methods, so the caller would need to test for `_min` and `min`...)

jorisvandenbossche commented:
> I think public is fine, as long as the API documentation makes it clear that directly using methods on Categorical or via ``s.values`` is prone to break over the time

Lets just make them internal then, no? Then we don't have that potential problem.

So to say I still vote for making them private functions. No user should ever use this I think? As we want them to use the categorical series, and not the raw Categorical. So a user should use `Series.dropna()`, and not `Categorical.dropna()` (there is also not a `ndarray.dropna()`). Providing it (even when documenting not to use it), will just tempt user to use it. 
And in an ideal world, it is still the goal that one day an array with a native `Categorical` type will be available (if not a numpy array, then maybe a dynd array), and in that case we want to replace `Categorical` with that I think (or at least the user facing `.values` categorical). So, until then, I would keep the public API of `Categorical` as much as close to a plain ndarray, and let the special functions it has be exposed for users through `.cat`

Kodiologist commented:
Note the changes to `test_categorical`. 0 counts and frequencies for unused categories seem more correct to me than NA counts and frequencies. Also, I removed one test entirely, which had a Categorical with both NA codes (-1) and an NA category, because that seems like a situation that shouldn't be allowed. We could enforce that rule, but I think that as things stand, the user needs to do some pretty silly things to get a Categorical like that, like renaming an existing level to `np.nan` while there are already NAs (-1 codes) in the Categorical, so enforcement may not be necessary.

Kodiologist commented:
@jreback, looks good?

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -20,6 +20,8 @@ users upgrade to this version.
 New features
 ~~~~~~~~~~~~
 
+- Added ``Categorical.value_counts`` and ``Categorical.dropna`` (:issue:`9443`)
```
take this out

jreback commented on pandas/core/algorithms.py:
```diff
@@ -211,47 +211,47 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
-    elif com.is_integer_dtype(dtype):
-        values = com._ensure_int64(values)
-        keys, counts = htable.value_count_int64(values)
+        if issubclass(values.dtype.type, (np.datetime64, np.timedelta64)) or is_period:
```
you can use ``is_datetimelike_array`` (which may not pass all tests), or ``is_datetime_or_timedelta_dtype`` which is a bit more strict (you can define this like ``is_period`` is above)

jreback commented on pandas/core/categorical.py:
```diff
@@ -881,6 +881,54 @@ def notnull(self):
+        keys, counts = htable.value_count_int64(com._ensure_int64(cat._codes))
+        result = Series(counts, index=keys)
+
+        ix = np.arange(len(cat.categories))
```
pass ``dtype='int64'`` here

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -1526,6 +1515,46 @@ def test_value_counts(self):
+    def test_value_counts_with_nan(self):
+        # https://github.com/pydata/pandas/issues/9443
+
+        s = pd.Series(["a", "b", "a"], dtype = "category")
```
when you are passing arguments, don't use spaces are the equals (e.g. pass as ``dropna=True``)

jreback commented on doc/source/whatsnew/v0.16.0.txt:
- Bug in ``Series.values_counts`` with excluding NaN for categorical type ``Series`` with ``dropna=True``
- ``Series.values_counts`` and ``Series.describe`` for categorical type will now put NaN entries at the end.

jreback commented:
@Kodiologist some minor comments, mostly stylist. Pls rebase and squash to a single commit. If you can do this today I wll put in 0.16.0

Kodiologist commented:
Sure thing.

Kodiologist committed:
BUG,TST,CLN: improve handling of Series.value_counts's argument 'dropna'

- Fixed bug in Series.values_counts with excluding NaN for categorical
  type Series with dropna=True.
- Series.values_counts and Series.describe for categorical type will now
  put NaN entries at the end.
- Series.describe for categorical type will now give counts and frequencies
  of 0, not NA, for unused categories.
- Added Categorical.value_counts and Categorical.dropna for internal use.

Kodiologist commented:
@jreback All done.

jreback commented:
shouldn't 3 exclude d?

```
In [1]: s = Series(pd.Categorical(['a', 'b', 'b', 'a','a', 'c', 'c', 'c'],categories=list('abcd')))

In [2]: s
Out[2]: 
0    a
1    b
2    b
3    a
4    a
5    c
6    c
7    c
dtype: category
Categories (4, object): [a < b < c < d]

In [3]: s.value_counts(dropna=True)
Out[3]: 
c    3
a    3
b    2
d    0
dtype: int64

In [4]: s.value_counts(dropna=False)
Out[4]: 
c    3
a    3
b    2
d    0
dtype: int64
```

jreback commented:
merged via https://github.com/pydata/pandas/commit/0270484ed9231349475d280526d9c58698c3475d

jreback commented:
@Kodiologist thanks!
